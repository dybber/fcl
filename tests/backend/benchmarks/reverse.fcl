sig reverse : forall 'a. ['a] -> ['a]
fun reverse arr =
  let n = lengthPull arr
  in generate n (fn i => index arr (n - i - 1))

sig simpleReverse : forall 'a. int -> ['a] -> Program <grid> ['a]<grid>
fun simpleReverse splitSize arr =
  splitUp splitSize arr
   |> mapPull (fn arr2 => return <block> (push <block> (reverse arr2)))
   |> reverse
   |> concat <block> splitSize


sig example : Program <grid> [int]<grid>
val example = simpleReverse 256 (generate (2048 * 2048) (fn i => i))

sig main : Program <grid> unit
val main =
  do <grid>
   { arr <- example
   ; forceAndPrint 10 arr
   ; benchmark 10 (bind <grid> example (force <grid>))
   }
