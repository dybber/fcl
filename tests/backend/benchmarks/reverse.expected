[4194303,4194302,4194301,4194300,4194299,4194298,4194297,4194296,4194295,4194294]
